### DISCUSSION TITLE
Can I use Colab as my Dev environment?

### DISCUSSION THREAD

**Discussion by mimoflynn** [microsoft/AI-Chat-App-Hack#30] [Q&A] [answered]
Hey MS Hackathon team, 
thanks for running this hackathon, excited to learn lots!
 
I've only ever used Google Colab as my Dev environment (using OpenAI APIs directly), so I'm not familiar with other dev environments e.g. codespaces, CLI, VScode. 

Is it possible to use a Colab notebook for this Hackathon, and if so, is there a sample notebook I can follow to set it up and connect to Azure? I want to build on Azure and I have an Azure account, with permission to create an 'Azure AI Studio' resource. 
Thanks
Mim

**Comment by pamelafox** [comment, marked as answer]
Hi @mimoflynn, thanks for joining. It will be difficult to create a full app with a frontend and backend inside a notebook environment, but you could certainly experiment with a RAG flow in a notebook. For example, here's a notebook that @khelanmodi showed in yesterday's CosmosDB talk:
https://github.com/microsoft/AzureDataRetrievalAugmentedGenerationSamples/blob/main/Python/CosmosDB-MongoDB-vCore/CosmosDB-MongoDB-vCore_AzureOpenAI_Tutorial.ipynb

I suggest trying that in a Colab notebook, and going from there.

**Reply by mimoflynn** [reply]
Thanks very much for the quick response. I'll check out the notebook and watch the Cosmos talk. 

One other question (answer may already be in that video I'm about to watch!) - my data is in a CSV file. Depending on the question, one or more columns in each row may need to be searched to find the best potential answers to send as RAG input to the LLM. Should I convert the CSV to a different format first e.g. text or jsonl or something before using it as a data source?

**Reply by pamelafox** [reply]
If your data is in CSV, then you can probably leave it that way, perhaps loaded in as a numpy array or pandas dataframe, if it's small enough to be all loaded into memory.

Then you can query the array/dataframe, get back N rows (3-10), and send those rows to OpenAI, along with their column headers. You may find better accuracy turning the rows into JSON at that point, actually, so that the column header is right next to the value. 

Another approach is to actually ask the LLM what query to run against the dataframe. I believe that's what's used by this notebook:
https://github.com/sdt-msft/Azure-Cognitive-Search-Azure-OpenAI-Accelerator/blob/main/06-TabularDataQA.ipynb

**Reply by mimoflynn** [reply]
@pamelafox I have followed along very closely with your video "Customizing your RAG Chat app" to see if I could try use Codespaces as Colab is very limited. I got the basic front and back end set up now for the azure-search-openai-demo, which is great. As part of setting up, I went connected the codespace to my Azure account, and created a workspace for it (I'm just about keeping up here!). However, when I checked on my Azure portal, there is now a new 'Unknown User' on there with 'Owner' permissions - see screenshot. Is this just a kind of 'virtual' owner representing Codespaces or something? I'm a little bit worried, so I'm hoping that's what it is!!  

![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/89992829/d0a8d079-e302-4df0-a14c-02beb9f164c9)

**Reply by pamelafox** [reply]
Hm, what part of the Portal shows that user? It's probably fine. How did you connect the Codespace to your Azure account, were you using the Azure Tools extension? 

For the repo, you only need to authenticate using azd, via `azd auth login`, though it's totally fine to authenticate with the extension if you want to take advantage of the tools in the Azure VS Code extension.

**Reply by mimoflynn** [reply]
Thanks very much for your response Pamela. So grateful for you checking in on the weekend!

**Connecting Codespace to Azure account**
To connect to my account, I clicked the Azure tab in the menu down the left. It told me to sign into my Azure account, and then the repo appeared in the 'Workspace' pane with an icon labelled 'Package provision and deploy'. I clicked on this and it started "Executing task: azd up", where I was asked for an env name - I used "chat-hack" and it then told me to run "azd auth login".  I ran it, it gave me a code (code changes, so it's ok to share it below in screenshot) and then asked me to confirm I was signing in to "Microsoft Azure CLI" (I got the gist of the message, and it seemed it was ok to say yes🤞). After that, it started "Executing task: azd up". It created a Resource Group rg-chat-hack on my Azure account. At that point, it seems to look like your video again! 

![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/89992829/d9e34c5b-b64d-4ebe-ad36-f8c20010853d)

![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/89992829/9756e77b-0815-44ba-a4a1-f8a6201a772a)

![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/89992829/8f7cbb9f-e3e0-4737-9acc-e8f741d0efa9)

![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/89992829/26d7fc8a-7d3c-4071-8c19-044e249adec9)
![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/89992829/ce133f81-7295-4d51-90fe-ca780f2d49fd)

**Portal area**
The part of the Portal that show that user is the Resource Group rg-chat-hack but it seems that User has been given 'Owner' permission on my Azure account, so it has access to everything else also. 

![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/89992829/3231f53b-2a4d-40a1-a7ce-8b06315014a2)

In some cases there are 2 unknown names e.g. under 'Contributors' below 
![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/89992829/2c2a01d7-9ab3-4b94-abf5-ee3a547d6664)
 
Thanks very much for your help in checking this out.
Mim

**Reply by pamelafox** [reply]
I did some poking around, and it does seem like an issue that happens occasionally for folks where some action created a user for you and then didn't clean up the user when they were no longer needed. (i.e. this blog post https://jloudon.com/cloud/Removing-Unknown-Azure-RBAC-Role-Assignments-with-PowerShell/) I think you could delete the users, as it looks like you've got the other roles required.

**Comment by mimoflynn** [comment]
Great Pamela.  Thanks so much for the follow up. I'll check out the blog
post 🙏🏻

**Reply by pallavigadepalli** [reply]
@mimoflynn  please let me know if you would like to connect on linkedin. We are facing something similar in our env and would love to hear about your experience. Thank you in advance. linkedin.com/in/pallavigadepalli. My devs use google colab and would like to use openAi APis. thank you!
